fuse_core/include/fuse_core/async_motion_model.h

```diff
@@ -171,13 +172,12 @@ class AsyncMotionModel : public MotionModel
   void stop() override;
 
 protected:
-  ros::CallbackQueue callback_queue_;  //!< The local callback queue used for all subscriptions
+  std::shared_ptr<fuse_core::CallbackAdapter> callback_queue_; //!< The callback queue used for fuse internal callbacks
   std::string name_;  //!< The unique name for this motion model instance
-  // rclcpp::Node::SharedPtr node_;  //!< The node for this motion model (TODO(CH3): Uncomment when it's time)
-  ros::NodeHandle node_handle_;  //!< A node handle in the global namespace using the local callback queue
-  ros::NodeHandle private_node_handle_;  //!< A node handle in the private namespace using the local callback queue
-  ros::AsyncSpinner spinner_;  //!< A single/multi-threaded spinner assigned to the local callback queue
-
+  rclcpp::Node::SharedPtr node_;  //!< The node for this motion model
```

Comment:
It looks like the node is unused except for adding the waitable to the executor.

Maybe we could pass the node or node interfaces [via `MotionModel::Initialize()`](https://github.com/methylDragon/fuse/blob/ee39c45cb7a704e0c646daf53993cfb5085f0d7c/fuse_core/include/fuse_core/motion_model.h#L97)? That eliminates the extra context and node, and it looks doable higher in the stack where [`MotionModel::initialize()` is called by an `Optimizer`](https://github.com/methylDragon/fuse/blob/ee39c45cb7a704e0c646daf53993cfb5085f0d7c/fuse_optimizers/src/optimizer.cpp#L182) which [does have a Node](https://github.com/methylDragon/fuse/blob/ee39c45cb7a704e0c646daf53993cfb5085f0d7c/fuse_optimizers/src/optimizer.cpp#L83).

Reply:
The node here is mostly used to structure parameters in the config files
It's reasonable to run the graph with dozens of named motion_models, and each one needs a unique name and parameters, each motion model can be associated with a different physical component that may or may not be included in any given deployment.

The separated node also allows complex motion_models (usually associated with a single robot chassis) to be targeted by ros cli tools as a separate entity in the ROS graph

Reply:
@BrettRD would you say that it's preferable then to keep the nodes separated? This was one of the places I planned to circle back on to try to combine it all into a single node (with the iffy bit being what happens when all the callback queues are merged), but the point you raised about CLI tools is pretty valid..

I'm not sure of the implications on lifecycle nodes and node components though.

Reply:
> The node here is mostly used to structure parameters in the config files
It's reasonable to run the graph with dozens of named motion_models, and each one needs a unique name and parameters, each motion model can be associated with a different physical component that may or may not be included in any given deployment.

I think we can still structure parameters in the config file by prepending `name_ + '.param_name'` to each parameter name.

> The separated node also allows complex motion_models (usually associated with a single robot chassis) to be targeted by ros cli tools as a separate entity in the ROS graph

I can think of a few reasons to avoid multiple nodes, but I'm not familiar enough with fuse to judge how useful it would be to a node per async motion model/publisher/sensor model. @svwilliams what are your thoughts?

Reasons to avoid multiple nodes:
* Nodes come with extra overhead like a graph guard condition, a rosout publisher, a parameter events publisher, and services for setting and getting parameters.
* Atomically changing parameters is scoped to a single node. A single node would allow someone to change parameters to two motion models atomically, such that if one failed (maybe from an invalid value) none of the changes would go through
* Building the nodes internally prevents choosing whether they use regular or lifecycle nodes.

Reply:
This is where my lack of experience with ROS2 is going to hinder me offering sound advice.

I can discuss the design intent in ROS1 and see if that helps.
* The motion models and sensors models act very node-like. They subscribe to sensor topics, process the data, and "publish" the results for consumption by the optimizer.
* The reason the sensor models and motion models are not stand-alone nodes or nodelets in ROS1 is the each model "publishes" a _derived class object_ instead of a message. Each model creates derived Constraints and Variables for use by the Optimizer. New user-derived models may create Constraints and Variables the Optimizer node cannot know about at compile-time. I could not think of an elegant solution to transmit derived objects using the standard pub-sub mechanisms in ROS1.
* Using ROS1 plugins, however, I can define an interface to the Optimizer that sends derived object pointers around, so that was what I implemented.
* But there is a downside to the plugin system. The Optimizer class creates the sensor and motion model instances as class variables. The Optimizer can call methods of the loaded plugins, so it would be easy to poll each loaded sensor and motion model for new data. But the sensor models and motion models are _producing_ data for the Optimizer. In an ideal world, the models would decide for themselves when they had new data and would push that into the Optimizer. It is difficult for the plugins to call methods of the Optimizer.
* To solve the polling/pushing issues, there are a series callbacks registered between the Optimizer and the Models. This results in rather complex threading implications.
* To hide all all of that complexity, the AsyncModels each run their own thread(s) and callback spinner. This makes them act a lot like a conventional ROS node or nodelet. Each derived Model can define the number of threads used to service its callbacks, it gets a private node handle in its namespace for parameter reading, etc.

Does any of that help?